File: swara-core/src/main/java/com/swara/core/composers/RandomComposer.java
package com.swara.core.composers;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.swara.core.Composer;
import com.swara.music.elements.Chord;
import com.swara.music.elements.Voice;

public class RandomComposer implements Composer {

    @Override
    public Stream<Chord> compose(List<Voice> examples) {
        // Aggregate all the various chords in all the example voices.
        final List<Chord> chords = examples.stream()
            .flatMap(voice -> voice.chords().stream())
            .collect(Collectors.toList());

        // Generate a stream of random chords sampled from the aggregated list.
        return Stream.generate(() -> chords.get((int) (Math.random() * chords.size())));
    }

}

File: swara-core/src/main/scala/com/swara/core/composers/RandomComposer.scala
package com.swara.core.composers

import com.swara.core.Composer
import com.swara.music.elements.{Chord, Voice}

import scala.util.Random
import collection.JavaConversions._

/**
 * A randomized composer. Random composers should produce chord sequences that are far less
 * musically correct and interesting than other methods; therefore, they can be used as a benchmark
 * to ensure that other techniques are working properly.
 */
class RandomComposer extends Composer {

  override def compose(examples: Seq[Voice]): Iterator[Chord] = {
    // Aggregate all the various chords in all the example voices.
    val chords = examples.flatMap(_.chords.toIndexedSeq)

    // Generate a stream of random chords sampled from the aggregated list.
    Iterator.continually(chords(Random.nextInt(chords.size)))
  }

}
